public/js/reducers/sources.js

```diff
@@ -156,24 +171,30 @@ function updateTabList(state, source, tabIndex) {
  * @memberof reducers/sources
  * @static
  */
-function getNewSelectedSourceId(state, id) : ?Source {
+function getNewSelectedSourceId(state, id: any) : ?Source {
```

Comment:
This logic is getting pretty hairy. It might be helpful to add some `close tabs` tests [here](https://github.com/devtools-html/debugger.html/blob/master/public/js/actions/tests/sources.js#L86).

Also, if you're interested in jumping into https://github.com/devtools-html/debugger.html/issues/804, I'd be happy to take a stab at some tests and perhaps a little clean up in this function :)